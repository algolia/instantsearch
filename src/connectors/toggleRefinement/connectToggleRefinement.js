import {
  checkRendering,
  escapeRefinement,
  unescapeRefinement,
} from '../../lib/utils.js';

import find from 'lodash/find';

const usage = `Usage:
var customToggle = connectToggleRefinement(function render(params, isFirstRendering) {
  // params = {
  //   value,
  //   createURL,
  //   refine,
  //   instantSearchInstance,
  //   widgetParams,
  // }
});
search.addWidget(
  customToggle({
    attribute,
    [on = true],
    [off],
  })
);
Full documentation available at https://community.algolia.com/instantsearch.js/v2/connectors/connectToggleRefinement.html
`;

/**
 * @typedef {Object} ToggleValue
 * @property {boolean} isRefined `true` if the toggle is on.
 * @property {number} count Number of results matched after applying the toggle refinement.
 * @property {Object} onFacetValue Value of the toggle when it's on.
 * @property {Object} offFacetValue Value of the toggle when it's off.
 */

/**
 * @typedef {Object} CustomToggleWidgetOptions
 * @property {string} attribute Name of the attribute for faceting (eg. "free_shipping").
 * @property {Object} [on=true] Value to filter on when toggled.
 * @property {Object} [off] Value to filter on when not toggled.
 */

/**
 * @typedef {Object} ToggleRenderingOptions
 * @property {ToggleValue} value The current toggle value.
 * @property {function(): string} createURL Creates an URL for the next state.
 * @property {function(value)} refine Updates to the next state by applying the toggle refinement.
 * @property {Object} widgetParams All original `CustomToggleWidgetOptions` forwarded to the `renderFn`.
 */

/**
 * **Toggle** connector provides the logic to build a custom widget that will provide
 * an on/off filtering feature based on an attribute value or values.
 *
 * Two modes are implemented in the custom widget:
 *  - with or without the value filtered
 *  - switch between two values.
 *
 * @type {Connector}
 * @param {function(ToggleRenderingOptions, boolean)} renderFn Rendering function for the custom **Toggle** widget.
 * @param {function} unmountFn Unmount function called when the widget is disposed.
 * @return {function(CustomToggleWidgetOptions)} Re-usable widget factory for a custom **Toggle** widget.
 * @example
 * // custom `renderFn` to render the custom ClearAll widget
 * function renderFn(ToggleRenderingOptions, isFirstRendering) {
 *   ToggleRenderingOptions.widgetParams.containerNode
 *     .find('a')
 *     .off('click');
 *
 *   var buttonHTML = `
 *     <a href="${ToggleRenderingOptions.createURL()}">
 *       <input
 *         type="checkbox"
 *         value="${ToggleRenderingOptions.value.name}"
 *         ${ToggleRenderingOptions.value.isRefined ? 'checked' : ''}
 *       />
 *       ${ToggleRenderingOptions.value.name} (${ToggleRenderingOptions.value.count})
 *     </a>
 *   `;
 *
 *   ToggleRenderingOptions.widgetParams.containerNode.html(buttonHTML);
 *   ToggleRenderingOptions.widgetParams.containerNode
 *     .find('a')
 *     .on('click', function(event) {
 *       event.preventDefault();
 *       event.stopPropagation();
 *
 *       ToggleRenderingOptions.refine(ToggleRenderingOptions.value);
 *     });
 * }
 *
 * // connect `renderFn` to Toggle logic
 * var customToggle = instantsearch.connectors.connectToggleRefinement(renderFn);
 *
 * // mount widget on the page
 * search.addWidget(
 *   customToggle({
 *     containerNode: $('#custom-toggle-container'),
 *     attribute: 'free_shipping',
 *   })
 * );
 */
export default function connectToggleRefinement(renderFn, unmountFn) {
  checkRendering(renderFn, usage);

  return (widgetParams = {}) => {
    const { attribute, on: userOn = true, off: userOff } = widgetParams;

    if (!attribute) {
      throw new Error(usage);
    }

    const hasAnOffValue = userOff !== undefined;
    const on = userOn ? escapeRefinement(userOn) : undefined;
    const off = userOff ? escapeRefinement(userOff) : undefined;

    return {
      getConfiguration() {
        return {
          disjunctiveFacets: [attribute],
        };
      },

      _toggleRefinement(helper, { isRefined } = {}) {
        // Checking
        if (!isRefined) {
          if (hasAnOffValue) {
            helper.removeDisjunctiveFacetRefinement(attribute, off);
          }
          helper.addDisjunctiveFacetRefinement(attribute, on);
        } else {
          // Unchecking
          helper.removeDisjunctiveFacetRefinement(attribute, on);
          if (hasAnOffValue) {
            helper.addDisjunctiveFacetRefinement(attribute, off);
          }
        }

        helper.search();
      },

      init({ state, helper, createURL, instantSearchInstance }) {
        this._createURL = isCurrentlyRefined => () =>
          createURL(
            state
              .removeDisjunctiveFacetRefinement(
                attribute,
                isCurrentlyRefined ? on : off
              )
              .addDisjunctiveFacetRefinement(
                attribute,
                isCurrentlyRefined ? off : on
              )
          );

        this.toggleRefinement = opts => {
          this._toggleRefinement(helper, opts);
        };

        const isRefined = state.isDisjunctiveFacetRefined(attribute, on);

        // no need to refine anything at init if no custom off values
        if (hasAnOffValue) {
          // Add filtering on the 'off' value if set
          if (!isRefined) {
            const currentPage = helper.getPage();
            helper
              .addDisjunctiveFacetRefinement(attribute, off)
              .setPage(currentPage);
          }
        }

        const onFacetValue = {
          isRefined,
          count: 0,
        };

        const offFacetValue = {
          isRefined: hasAnOffValue && !isRefined,
          count: 0,
        };

        const value = {
          isRefined,
          count: null,
          onFacetValue,
          offFacetValue,
        };

        renderFn(
          {
            value,
            createURL: this._createURL(value.isRefined),
            refine: this.toggleRefinement,
            instantSearchInstance,
            widgetParams,
          },
          true
        );
      },

      render({ helper, results, state, instantSearchInstance }) {
        const isRefined = helper.state.isDisjunctiveFacetRefined(attribute, on);
        const offValue = off === undefined ? false : off;
        const allFacetValues = results.getFacetValues(attribute);

        const onData = find(
          allFacetValues,
          ({ name }) => name === unescapeRefinement(on)
        );
        const onFacetValue = {
          isRefined: onData !== undefined ? onData.isRefined : false,
          count: onData === undefined ? null : onData.count,
        };

        const offData = hasAnOffValue
          ? find(
              allFacetValues,
              ({ name }) => name === unescapeRefinement(offValue)
            )
          : undefined;
        const offFacetValue = {
          isRefined: offData !== undefined ? offData.isRefined : false,
          count:
            offData === undefined
              ? allFacetValues.reduce((total, { count }) => total + count, 0)
              : offData.count,
        };

        // what will we show by default,
        // if checkbox is not checked, show: [ ] free shipping (countWhenChecked)
        // if checkbox is checked, show: [x] free shipping (countWhenNotChecked)
        const nextRefinement = isRefined ? offFacetValue : onFacetValue;

        const value = {
          isRefined,
          count: nextRefinement === undefined ? null : nextRefinement.count,
          onFacetValue,
          offFacetValue,
        };

        renderFn(
          {
            value,
            state,
            createURL: this._createURL(value.isRefined),
            refine: this.toggleRefinement,
            helper,
            instantSearchInstance,
            widgetParams,
          },
          false
        );
      },

      dispose({ state }) {
        unmountFn();

        const nextState = state
          .removeDisjunctiveFacetRefinement(attribute)
          .removeDisjunctiveFacet(attribute);

        return nextState;
      },

      getWidgetState(uiState, { searchParameters }) {
        const isRefined = searchParameters.isDisjunctiveFacetRefined(
          attribute,
          on
        );

        if (
          !isRefined ||
          (uiState && uiState.toggle && uiState.toggle[attribute] === isRefined)
        ) {
          return uiState;
        }

        return {
          ...uiState,
          toggle: {
            ...uiState.toggle,
            [attribute]: isRefined,
          },
        };
      },

      getWidgetSearchParameters(searchParameters, { uiState }) {
        const isRefined = Boolean(uiState.toggle && uiState.toggle[attribute]);

        if (isRefined) {
          if (hasAnOffValue)
            return searchParameters
              .removeDisjunctiveFacetRefinement(attribute, off)
              .addDisjunctiveFacetRefinement(attribute, on);

          return searchParameters.addDisjunctiveFacetRefinement(attribute, on);
        }

        if (hasAnOffValue)
          return searchParameters
            .removeDisjunctiveFacetRefinement(attribute, on)
            .addDisjunctiveFacetRefinement(attribute, off);

        return searchParameters.removeDisjunctiveFacetRefinement(attribute, on);
      },
    };
  };
}
