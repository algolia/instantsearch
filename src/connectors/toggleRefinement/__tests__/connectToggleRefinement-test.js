import jsHelper, {
  SearchResults,
  SearchParameters,
} from 'algoliasearch-helper';
import connectToggleRefinement from '../connectToggleRefinement';

describe('connectToggleRefinement', () => {
  describe('Usage', () => {
    it('throws without render function', () => {
      expect(() => {
        connectToggleRefinement()({});
      }).toThrowErrorMatchingInlineSnapshot(`
"The render function is not valid (got type \\"undefined\\").

See documentation: https://www.algolia.com/doc/api-reference/widgets/toggle-refinement/js/#connector"
`);
    });

    it('throws without attribute option', () => {
      expect(() => {
        connectToggleRefinement(() => {})({});
      }).toThrowErrorMatchingInlineSnapshot(`
"The \`attribute\` option is required.

See documentation: https://www.algolia.com/doc/api-reference/widgets/toggle-refinement/js/#connector"
`);
    });

    it('is a widget', () => {
      const render = jest.fn();
      const unmount = jest.fn();

      const customToggleRefinement = connectToggleRefinement(render, unmount);
      const widget = customToggleRefinement({ attribute: 'facet' });

      expect(widget).toEqual(
        expect.objectContaining({
          $$type: 'ais.toggleRefinement',
          init: expect.any(Function),
          render: expect.any(Function),
          dispose: expect.any(Function),
          getConfiguration: expect.any(Function),
          getWidgetState: expect.any(Function),
          getWidgetSearchParameters: expect.any(Function),
        })
      );
    });
  });

  it('Renders during init and render', () => {
    // test that the dummyRendering is called with the isFirstRendering
    // flag set accordingly
    const rendering = jest.fn();
    const makeWidget = connectToggleRefinement(rendering);

    const attribute = 'isShippingFree';
    const widget = makeWidget({
      attribute,
    });

    const config = widget.getConfiguration();
    expect(config).toEqual({
      disjunctiveFacets: [attribute],
    });

    const helper = jsHelper({}, '', config);
    helper.search = jest.fn();

    widget.init({
      helper,
      state: helper.state,
      createURL: () => '#',
    });

    {
      // should call the rendering once with isFirstRendering to true
      expect(rendering).toHaveBeenCalledTimes(1);
      const isFirstRendering =
        rendering.mock.calls[rendering.mock.calls.length - 1][1];
      expect(isFirstRendering).toBe(true);

      // should provide good values for the first rendering
      const { value, widgetParams } = rendering.mock.calls[
        rendering.mock.calls.length - 1
      ][0];
      expect(value).toEqual({
        name: 'isShippingFree',
        count: null,
        isRefined: false,
        onFacetValue: {
          isRefined: false,
          count: 0,
        },
        offFacetValue: {
          isRefined: false,
          count: 0,
        },
      });

      expect(widgetParams).toEqual({
        attribute,
      });
    }

    widget.render({
      results: new SearchResults(helper.state, [
        {
          facets: {
            isShippingFree: {
              true: 45,
              false: 40,
            },
          },
          nbHits: 85,
        },
      ]),
      state: helper.state,
      helper,
      createURL: () => '#',
    });

    {
      // Should call the rendering a second time, with isFirstRendering to false
      expect(rendering).toHaveBeenCalledTimes(2);
      const isFirstRendering =
        rendering.mock.calls[rendering.mock.calls.length - 1][1];
      expect(isFirstRendering).toBe(false);

      // should provide good values after the first search
      const { value } = rendering.mock.calls[
        rendering.mock.calls.length - 1
      ][0];
      expect(value).toEqual({
        name: 'isShippingFree',
        count: 45,
        isRefined: false,
        onFacetValue: {
          isRefined: false,
          count: 45,
        },
        offFacetValue: {
          isRefined: false,
          count: 85,
        },
      });
    }
  });

  it('Provides a function to add/remove a facet value', () => {
    const rendering = jest.fn();
    const makeWidget = connectToggleRefinement(rendering);

    const attribute = 'isShippingFree';
    const widget = makeWidget({
      attribute,
    });

    const helper = jsHelper({}, '', widget.getConfiguration());
    helper.search = jest.fn();

    widget.init({
      helper,
      state: helper.state,
      createURL: () => '#',
    });

    {
      // first rendering
      expect(helper.state.disjunctiveFacetsRefinements[attribute]).toEqual(
        undefined
      );
      const renderOptions =
        rendering.mock.calls[rendering.mock.calls.length - 1][0];
      const { refine, value } = renderOptions;
      expect(value).toEqual({
        name: 'isShippingFree',
        count: null,
        isRefined: false,
        onFacetValue: {
          isRefined: false,
          count: 0,
        },
        offFacetValue: {
          isRefined: false,
          count: 0,
        },
      });
      refine({ isRefined: value.isRefined });
      expect(helper.state.disjunctiveFacetsRefinements[attribute]).toEqual([
        'true',
      ]);
      refine({ isRefined: !value.isRefined });
      expect(helper.state.disjunctiveFacetsRefinements[attribute]).toEqual([]);
    }

    widget.render({
      results: new SearchResults(helper.state, [
        {
          facets: {
            isShippingFree: {
              true: 45,
              false: 40,
            },
          },
          nbHits: 85,
        },
      ]),
      state: helper.state,
      helper,
      createURL: () => '#',
    });

    {
      // Second rendering
      expect(helper.state.disjunctiveFacetsRefinements[attribute]).toEqual([]);
      const renderOptions =
        rendering.mock.calls[rendering.mock.calls.length - 1][0];
      const { refine, value } = renderOptions;
      expect(value).toEqual({
        name: 'isShippingFree',
        count: 45,
        isRefined: false,
        onFacetValue: {
          isRefined: false,
          count: 45,
        },
        offFacetValue: {
          isRefined: false,
          count: 85,
        },
      });
      refine({ isRefined: value.isRefined });
      expect(helper.state.disjunctiveFacetsRefinements[attribute]).toEqual([
        'true',
      ]);
    }

    widget.render({
      results: new SearchResults(helper.state, [
        {
          facets: {
            isShippingFree: {
              true: 45,
            },
          },
          nbHits: 85,
        },
        {
          facets: {
            isShippingFree: {
              true: 45,
              false: 40,
            },
          },
          nbHits: 85,
        },
      ]),
      state: helper.state,
      helper,
      createURL: () => '#',
    });

    {
      // Third rendering
      expect(helper.state.disjunctiveFacetsRefinements[attribute]).toEqual([
        'true',
      ]);
      const renderOptions =
        rendering.mock.calls[rendering.mock.calls.length - 1][0];
      const { refine, value } = renderOptions;
      expect(value).toEqual({
        name: 'isShippingFree',
        count: 85,
        isRefined: true,
        onFacetValue: {
          isRefined: true,
          count: 45,
        },
        offFacetValue: {
          isRefined: false,
          count: 85,
        },
      });
      refine(value);
      expect(helper.state.disjunctiveFacetsRefinements[attribute]).toEqual([]);
    }
  });

  it('Provides a function to toggle between two values', () => {
    const rendering = jest.fn();
    const makeWidget = connectToggleRefinement(rendering);

    const attribute = 'isShippingFree';
    const widget = makeWidget({
      attribute,
      on: 'true',
      off: 'false',
    });

    const helper = jsHelper({}, '', widget.getConfiguration());
    helper.search = jest.fn();

    widget.init({
      helper,
      state: helper.state,
      createURL: () => '#',
    });

    {
      // first rendering
      expect(helper.state.disjunctiveFacetsRefinements[attribute]).toEqual([
        'false',
      ]);
      const renderOptions =
        rendering.mock.calls[rendering.mock.calls.length - 1][0];
      const { refine, value } = renderOptions;

      expect(value).toEqual({
        name: 'isShippingFree',
        count: null,
        isRefined: false,
        onFacetValue: {
          isRefined: false,
          count: 0,
        },
        offFacetValue: {
          isRefined: true,
          count: 0,
        },
      });
      refine({ isRefined: value.isRefined });
      expect(helper.state.disjunctiveFacetsRefinements[attribute]).toEqual([
        'true',
      ]);
      refine({ isRefined: !value.isRefined });
      expect(helper.state.disjunctiveFacetsRefinements[attribute]).toEqual([
        'false',
      ]);
    }

    widget.render({
      results: new SearchResults(helper.state, [
        {
          facets: {
            isShippingFree: {
              false: 40,
            },
          },
          nbHits: 40,
        },
        {
          facets: {
            isShippingFree: {
              true: 45,
              false: 40,
            },
          },
          nbHits: 85,
        },
      ]),
      state: helper.state,
      helper,
      createURL: () => '#',
    });

    {
      // Second rendering
      expect(helper.state.disjunctiveFacetsRefinements[attribute]).toEqual([
        'false',
      ]);
      const renderOptions =
        rendering.mock.calls[rendering.mock.calls.length - 1][0];
      const { refine, value } = renderOptions;
      expect(value).toEqual({
        // the value is the one that is not selected
        name: 'isShippingFree',
        count: 45,
        isRefined: false,
        onFacetValue: {
          isRefined: false,
          count: 45,
        },
        offFacetValue: {
          isRefined: true,
          count: 40,
        },
      });
      refine({ isRefined: value.isRefined });
      expect(helper.state.disjunctiveFacetsRefinements[attribute]).toEqual([
        'true',
      ]);
    }

    widget.render({
      results: new SearchResults(helper.state, [
        {
          facets: {
            isShippingFree: {
              true: 45,
            },
          },
          nbHits: 85,
        },
        {
          facets: {
            isShippingFree: {
              true: 45,
              false: 40,
            },
          },
          nbHits: 85,
        },
      ]),
      state: helper.state,
      helper,
      createURL: () => '#',
    });

    {
      // Third rendering
      expect(helper.state.disjunctiveFacetsRefinements[attribute]).toEqual([
        'true',
      ]);
      const renderOptions =
        rendering.mock.calls[rendering.mock.calls.length - 1][0];
      const { refine, value } = renderOptions;
      expect(value).toEqual({
        name: 'isShippingFree',
        count: 40,
        isRefined: true,
        onFacetValue: {
          isRefined: true,
          count: 45,
        },
        offFacetValue: {
          isRefined: false,
          count: 40,
        },
      });
      refine({ isRefined: value.isRefined });
      expect(helper.state.disjunctiveFacetsRefinements[attribute]).toEqual([
        'false',
      ]);
    }
  });

  it('sets user-provided "off" value by default (falsy)', () => {
    const makeWidget = connectToggleRefinement(() => {});
    const widget = makeWidget({
      attribute: 'whatever',
      off: false,
    });

    const helper = jsHelper({}, '', widget.getConfiguration());
    widget.init({ helper, state: helper.state });

    expect(helper.state.disjunctiveFacetsRefinements).toEqual(
      expect.objectContaining({
        whatever: ['false'],
      })
    );
  });

  it('sets user-provided "off" value by default (truthy)', () => {
    const makeWidget = connectToggleRefinement(() => {});
    const widget = makeWidget({
      attribute: 'whatever',
      off: true,
    });

    const helper = jsHelper({}, '', widget.getConfiguration());
    widget.init({ helper, state: helper.state });

    expect(helper.state.disjunctiveFacetsRefinements).toEqual(
      expect.objectContaining({
        whatever: ['true'],
      })
    );
  });

  it('sets user-provided "on" value on refine (falsy)', () => {
    let caughtRefine;
    const makeWidget = connectToggleRefinement(({ refine }) => {
      caughtRefine = refine;
    });
    const widget = makeWidget({
      attribute: 'whatever',
      on: false,
    });

    const helper = jsHelper({ search() {} }, '', widget.getConfiguration());
    helper.search = jest.fn();

    widget.init({ helper, state: helper.state });

    expect(helper.state.disjunctiveFacetsRefinements).toEqual({});

    widget.render({
      results: new SearchResults(helper.state, [
        {
          facets: {
            whatever: {
              true: 45,
              false: 40,
            },
          },
          nbHits: 85,
        },
      ]),
      state: helper.state,
      helper,
    });

    // toggle the value
    caughtRefine();

    expect(helper.state.disjunctiveFacetsRefinements).toEqual(
      expect.objectContaining({
        whatever: ['false'],
      })
    );
  });

  describe('routing', () => {
    const getInitializedWidget = (config = {}) => {
      const rendering = jest.fn();
      const makeWidget = connectToggleRefinement(rendering);

      const attribute = 'isShippingFree';
      const widget = makeWidget({
        attribute,
        ...config,
      });

      const initialConfig = widget.getConfiguration({});
      const helper = jsHelper({}, '', initialConfig);
      helper.search = jest.fn();

      widget.init({
        helper,
        state: helper.state,
        createURL: () => '#',
      });

      const { refine } = rendering.mock.calls[
        rendering.mock.calls.length - 1
      ][0];

      return [widget, helper, refine];
    };

    describe('getWidgetState', () => {
      test('should give back the object unmodified if the default value is selected', () => {
        const [widget, helper] = getInitializedWidget();
        const uiStateBefore = {};
        const uiStateAfter = widget.getWidgetState(uiStateBefore, {
          searchParameters: helper.state,
          helper,
        });
        expect(uiStateAfter).toBe(uiStateBefore);
      });

      test('should add an entry equal to the refinement', () => {
        const [widget, helper, refine] = getInitializedWidget();
        refine({ isRefined: false }); // refinement is based on the previous value (the one passed)
        const uiStateBefore = {};
        const uiStateAfter = widget.getWidgetState(uiStateBefore, {
          searchParameters: helper.state,
          helper,
        });
        expect(uiStateAfter).toMatchSnapshot();
      });

      test('should return the same UI state if the value if the refinement is the same', () => {
        const [widget, helper, refine] = getInitializedWidget();
        refine({ isRefined: false });
        const uiStateBefore = widget.getWidgetState(
          {},
          {
            searchParameters: helper.state,
            helper,
          }
        );
        const uiStateAfter = widget.getWidgetState(uiStateBefore, {
          searchParameters: helper.state,
          helper,
        });
        expect(uiStateAfter).toBe(uiStateBefore);
      });
    });

    describe('getWidgetSearchParameters', () => {
      test('should return the same SP if no value is in the UI state (one value)', () => {
        const [widget, helper] = getInitializedWidget();
        const uiState = {};
        const searchParametersBefore = SearchParameters.make(helper.state);
        const searchParametersAfter = widget.getWidgetSearchParameters(
          searchParametersBefore,
          { uiState }
        );
        expect(searchParametersAfter).toBe(searchParametersBefore);
      });

      test('should enforce the default value if no value is in the UI state (two values)', () => {
        const [widget, helper] = getInitializedWidget({
          on: 'free-shipping',
          off: 'paid-shipping',
        });
        const uiState = {};
        const searchParametersBefore = SearchParameters.make(helper.state);
        const searchParametersAfter = widget.getWidgetSearchParameters(
          searchParametersBefore,
          { uiState }
        );
        expect(searchParametersAfter).toBe(searchParametersBefore);
      });

      test('should remove the refinement (one value)', () => {
        const [widget, helper, refine] = getInitializedWidget();
        refine({ isRefined: false });
        const uiState = {};
        const searchParametersBefore = SearchParameters.make(helper.state);
        const searchParametersAfter = widget.getWidgetSearchParameters(
          searchParametersBefore,
          { uiState }
        );
        expect(searchParametersAfter).toMatchSnapshot();
      });

      test('should update the SP base on the UI state (two values)', () => {
        const [widget, helper, refine] = getInitializedWidget({
          on: 'free-shipping',
          off: 'paid-shipping',
        });
        refine({ isRefined: false });
        const uiState = {};
        const searchParametersBefore = SearchParameters.make(helper.state);
        const searchParametersAfter = widget.getWidgetSearchParameters(
          searchParametersBefore,
          { uiState }
        );
        expect(searchParametersAfter).toMatchSnapshot();
      });

      test('should update the SP base on the UI state - toggled (two values)', () => {
        const [widget, helper] = getInitializedWidget({
          on: 'free-shipping',
          off: 'paid-shipping',
        });
        const uiState = {
          toggle: {
            isShippingFree: true,
          },
        };
        const searchParametersBefore = SearchParameters.make(helper.state);
        const searchParametersAfter = widget.getWidgetSearchParameters(
          searchParametersBefore,
          { uiState }
        );
        expect(searchParametersAfter).toMatchSnapshot();
      });
    });
  });
});
