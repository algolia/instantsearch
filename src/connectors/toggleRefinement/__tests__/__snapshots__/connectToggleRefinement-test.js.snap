// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`connectToggleRefinement routing getWidgetSearchParameters should remove the refinement (one value) 1`] = `
SearchParameters {
  "advancedSyntax": undefined,
  "allowTyposOnNumericTokens": undefined,
  "analytics": undefined,
  "analyticsTags": undefined,
  "aroundLatLng": undefined,
  "aroundLatLngViaIP": undefined,
  "aroundPrecision": undefined,
  "aroundRadius": undefined,
  "attributesToHighlight": undefined,
  "attributesToRetrieve": undefined,
  "attributesToSnippet": undefined,
  "disableExactOnAttributes": undefined,
  "disjunctiveFacets": Array [
    "isShippingFree",
  ],
  "disjunctiveFacetsRefinements": Object {},
  "distinct": undefined,
  "enableExactOnSingleWordQuery": undefined,
  "facets": Array [],
  "facetsExcludes": Object {},
  "facetsRefinements": Object {},
  "getRankingInfo": undefined,
  "hierarchicalFacets": Array [],
  "hierarchicalFacetsRefinements": Object {},
  "highlightPostTag": undefined,
  "highlightPreTag": undefined,
  "hitsPerPage": undefined,
  "ignorePlurals": undefined,
  "index": "",
  "insideBoundingBox": undefined,
  "insidePolygon": undefined,
  "length": undefined,
  "maxValuesPerFacet": undefined,
  "minProximity": undefined,
  "minWordSizefor1Typo": undefined,
  "minWordSizefor2Typos": undefined,
  "minimumAroundRadius": undefined,
  "numericFilters": undefined,
  "numericRefinements": Object {},
  "offset": undefined,
  "optionalFacetFilters": undefined,
  "optionalTagFilters": undefined,
  "optionalWords": undefined,
  "page": 0,
  "query": "",
  "queryType": undefined,
  "removeWordsIfNoResults": undefined,
  "replaceSynonymsInHighlight": undefined,
  "restrictSearchableAttributes": undefined,
  "snippetEllipsisText": undefined,
  "synonyms": undefined,
  "tagFilters": undefined,
  "tagRefinements": Array [],
  "typoTolerance": undefined,
}
`;

exports[`connectToggleRefinement routing getWidgetSearchParameters should update the SP base on the UI state (two values) 1`] = `
SearchParameters {
  "advancedSyntax": undefined,
  "allowTyposOnNumericTokens": undefined,
  "analytics": undefined,
  "analyticsTags": undefined,
  "aroundLatLng": undefined,
  "aroundLatLngViaIP": undefined,
  "aroundPrecision": undefined,
  "aroundRadius": undefined,
  "attributesToHighlight": undefined,
  "attributesToRetrieve": undefined,
  "attributesToSnippet": undefined,
  "disableExactOnAttributes": undefined,
  "disjunctiveFacets": Array [
    "isShippingFree",
  ],
  "disjunctiveFacetsRefinements": Object {
    "isShippingFree": Array [
      "paid-shipping",
    ],
  },
  "distinct": undefined,
  "enableExactOnSingleWordQuery": undefined,
  "facets": Array [],
  "facetsExcludes": Object {},
  "facetsRefinements": Object {},
  "getRankingInfo": undefined,
  "hierarchicalFacets": Array [],
  "hierarchicalFacetsRefinements": Object {},
  "highlightPostTag": undefined,
  "highlightPreTag": undefined,
  "hitsPerPage": undefined,
  "ignorePlurals": undefined,
  "index": "",
  "insideBoundingBox": undefined,
  "insidePolygon": undefined,
  "length": undefined,
  "maxValuesPerFacet": undefined,
  "minProximity": undefined,
  "minWordSizefor1Typo": undefined,
  "minWordSizefor2Typos": undefined,
  "minimumAroundRadius": undefined,
  "numericFilters": undefined,
  "numericRefinements": Object {},
  "offset": undefined,
  "optionalFacetFilters": undefined,
  "optionalTagFilters": undefined,
  "optionalWords": undefined,
  "page": 0,
  "query": "",
  "queryType": undefined,
  "removeWordsIfNoResults": undefined,
  "replaceSynonymsInHighlight": undefined,
  "restrictSearchableAttributes": undefined,
  "snippetEllipsisText": undefined,
  "synonyms": undefined,
  "tagFilters": undefined,
  "tagRefinements": Array [],
  "typoTolerance": undefined,
}
`;

exports[`connectToggleRefinement routing getWidgetSearchParameters should update the SP base on the UI state - toggled (two values) 1`] = `
SearchParameters {
  "advancedSyntax": undefined,
  "allowTyposOnNumericTokens": undefined,
  "analytics": undefined,
  "analyticsTags": undefined,
  "aroundLatLng": undefined,
  "aroundLatLngViaIP": undefined,
  "aroundPrecision": undefined,
  "aroundRadius": undefined,
  "attributesToHighlight": undefined,
  "attributesToRetrieve": undefined,
  "attributesToSnippet": undefined,
  "disableExactOnAttributes": undefined,
  "disjunctiveFacets": Array [
    "isShippingFree",
  ],
  "disjunctiveFacetsRefinements": Object {
    "isShippingFree": Array [
      "free-shipping",
    ],
  },
  "distinct": undefined,
  "enableExactOnSingleWordQuery": undefined,
  "facets": Array [],
  "facetsExcludes": Object {},
  "facetsRefinements": Object {},
  "getRankingInfo": undefined,
  "hierarchicalFacets": Array [],
  "hierarchicalFacetsRefinements": Object {},
  "highlightPostTag": undefined,
  "highlightPreTag": undefined,
  "hitsPerPage": undefined,
  "ignorePlurals": undefined,
  "index": "",
  "insideBoundingBox": undefined,
  "insidePolygon": undefined,
  "length": undefined,
  "maxValuesPerFacet": undefined,
  "minProximity": undefined,
  "minWordSizefor1Typo": undefined,
  "minWordSizefor2Typos": undefined,
  "minimumAroundRadius": undefined,
  "numericFilters": undefined,
  "numericRefinements": Object {},
  "offset": undefined,
  "optionalFacetFilters": undefined,
  "optionalTagFilters": undefined,
  "optionalWords": undefined,
  "page": 0,
  "query": "",
  "queryType": undefined,
  "removeWordsIfNoResults": undefined,
  "replaceSynonymsInHighlight": undefined,
  "restrictSearchableAttributes": undefined,
  "snippetEllipsisText": undefined,
  "synonyms": undefined,
  "tagFilters": undefined,
  "tagRefinements": Array [],
  "typoTolerance": undefined,
}
`;

exports[`connectToggleRefinement routing getWidgetState should add an entry equal to the refinement 1`] = `
Object {
  "toggle": Object {
    "isShippingFree": true,
  },
}
`;
